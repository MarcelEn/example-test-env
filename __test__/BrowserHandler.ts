import puppeteer, { Browser, Page } from "puppeteer";

const target = "http://localhost:3000";

export class PageHandler {
    constructor(public page: Page) { }

    public matchImage = async () => {
        for (const [device, width] of Object.entries({
            mobile: 580,
            tablet: 1050,
            desktop: 1920,
        })) {
            await this.page.setViewport(
                {
                    ...this.page.viewport(),
                    width,
                }
            );
            const height = await this.page.$eval(".content", (e) => (e as any).offsetHeight);

            await this.page.setViewport(
                {
                    ...this.page.viewport(),
                    height,
                }
            );

            await this.page.evaluate(() => {
                document.querySelector(".content").scrollIntoView();
            });

            (expect(await this.page.screenshot()) as any)
                .toMatchImageSnapshot({
                    customSnapshotIdentifier: ({ defaultIdentifier }) => {
                        return `${defaultIdentifier.slice(13, defaultIdentifier.length)}-${device}`
                    }
                });
        }
    }

    public matchSnapshot = async () => {
        let snapshot = await this.page.$eval(".content", ({ innerHTML }) => innerHTML)
            .catch(e => "");

        expect(snapshot).toMatchSnapshot();
    }

    public close = async () => {
        await this.page.close();
    }
}

export default class BrowserHandler {
    private browser: Browser;

    public init = async () => {
        if (!!this.browser) return;
        this.browser = await puppeteer.launch({
            // headless: false,
            // slowMo: 100,
        });
    }

    public createPageHandler = async (data: string) => {
        const page = await this.browser.newPage();

        await page.goto(target);
        await page.evaluate((innerHTML) => {
            document.querySelector("textarea").innerHTML = innerHTML;
            document.querySelector("form").submit();
        }, data);

        await page.waitForNavigation({ waitUntil: 'networkidle0' });

        return new PageHandler(page);
    };

    public close = async () => {
        await Promise.all((await this.browser.pages()).map(page => page.close()));
        await this.browser.close();
    }
}