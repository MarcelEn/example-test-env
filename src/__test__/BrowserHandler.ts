import puppeteer, { Browser, Page } from "puppeteer";
import { toMatchImageSnapshot } from "jest-image-snapshot";

expect.extend({ toMatchImageSnapshot });
jest.setTimeout(60000);

const headless = false//process.env.HEADLESS !== "false";

export class PageHandler {
    constructor(public page: Page) { }

    public async matchImage() {
        for (const [device, width] of Object.entries({
            mobile: 580,
            tablet: 1050,
            desktop: 1920,
        })) {
            await this.page.setViewport(
                {
                    ...this.page.viewport(),
                    width,
                }
            );
            const height = await this.page.$eval("#content", (e) => (e as any).offsetHeight);

            await this.page.setViewport(
                {
                    ...this.page.viewport(),
                    height,
                }
            );

            await this.page.evaluate(() => {
                document.querySelector("#content").scrollIntoView();
            });

            (expect(await this.page.screenshot()) as any)
                .toMatchImageSnapshot({
                    failureThreshold: headless ? 0 : 1,
                    failureThresholdType: 'percent',
                    customSnapshotIdentifier: ({ defaultIdentifier }) => `${defaultIdentifier}-${device}`
                });
        }
    }

    public async matchSnapshot() {
        let snapshot = await this.page.$eval("#content", ({ innerHTML }) => innerHTML)
            .catch(e => "");

        expect(snapshot).toMatchSnapshot();
    }

    public async close() {
        await this.page.close();
    }
}

export class BrowserHandler {
    private browser: Browser;

    public async init() {
        this.browser = await puppeteer.launch({
            headless,
            slowMo: headless ? 0 : 100,
            ignoreDefaultArgs: ["--hide-scrollbars"]
        });
    }

    public async createPageHandler(data: string) {
        const page = await this.browser.newPage();

        await page.goto("http://localhost:3000/form.html");
        await page.evaluate((innerHTML) => {
            document.querySelector("textarea").innerHTML = innerHTML;
            document.querySelector("form").submit();
        }, data);

        await page.waitForNavigation({ waitUntil: 'networkidle0' });

        await page.evaluate(() => {
            var consentConfirm = document.getElementById("selectAll");
            if (!!consentConfirm) consentConfirm.click();

            var stageassistent = document.getElementById("stageassistent");
            if (!!stageassistent) stageassistent.remove();

            document.querySelector("html").setAttribute("style", "overflow: hidden");
        });

        return new PageHandler(page);
    };

    public async close() {
        await this.browser.close();
    }
}

export default new BrowserHandler();