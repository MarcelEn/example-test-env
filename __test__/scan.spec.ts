import fs from "fs";
import BrowserHandler from "./BrowserHandler";
import { toMatchImageSnapshot } from "jest-image-snapshot";

expect.extend({ toMatchImageSnapshot });

const componentPath = "__test__/components";

interface TestSuite {
    [describer: string]: {
        [itter: string]: {
            data: string;
            additionalProcess?: string;
        }
    }
}

const testSuite: TestSuite = {};

fs.readdirSync(componentPath)
    .map(describer => ({
        describer,
        itters: fs.readdirSync(`${componentPath}/${describer}`),
    }))
    .map(({ describer, itters }) => {
        testSuite[describer] = {};
        itters.forEach(itter => {
            testSuite[describer][itter] = {
                data: fs.readFileSync(`${componentPath}/${describer}/${itter}/data.xml`).toString(),
                additionalProcess: fs.existsSync(`${componentPath}/${describer}/${itter}/process.ts`) ?
                    `./components/${describer}/${itter}/process.ts`
                    :
                    undefined
            };
        });
    })


jest.setTimeout(30000);

let browserHandler: BrowserHandler = new BrowserHandler();

const testSuiteCount = Object.entries(testSuite).map(([_, entry]) => entry).map(obj => Object.keys(obj)).reduce((p = [], c) => [...p, ...c]).length;
let testSuitesDone = 0;

const checkBrowserClose = async () => {
    if (testSuitesDone >= testSuiteCount) {
        await browserHandler.close();
    }
}

Object.keys(testSuite).forEach(describer => {
    describe(describer, () => {
        Object.keys(testSuite[describer]).forEach(itter => {
            const { data, additionalProcess } = testSuite[describer][itter];
            it(itter, async (done) => {
                try {
                    await browserHandler.init();
                    const pageHandler = await browserHandler.createPageHandler(data);
                    await pageHandler.matchSnapshot();
                    await pageHandler.matchImage();

                    if (!!additionalProcess) {
                        await (await import(additionalProcess)).default(pageHandler);
                    }
                    done();
                } finally {
                    testSuitesDone++;
                    await checkBrowserClose();
                }
            });
        });
    });
});