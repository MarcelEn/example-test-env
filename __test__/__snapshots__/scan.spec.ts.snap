// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kickass-stage renders-correctly 1`] = `
"
            <h1>Kickass Stage</h1>

        "
`;

exports[`triple-stage renders-correctly 1`] = `
"
            <h1>Triple Stage</h1>

        "
`;

exports[`triple-stage renders-correctly 5`] = `
"
            <h1>nope</h1>

        "
`;
